### ISSUE TITLE
[v4] importing tailwindcss/utilities breaks hmr

### ISSUE THREAD

**Issue by Matheun** [tailwindlabs/tailwindcss#14020]
Im using:
- tailwind v4.0.0-alpha.17 ( @tailwindcss/postcss )
- nuxt v3.12.3
- sass v1.77.8

The following code works perfectly:
```
@import "../theme/_index.scss" layer(theme);

@import "tailwindcss/preflight" layer(base);
@import "../base/_index.scss" layer(base);
@import "../plugins/postcss/_index.scss" layer(base);

@import "../plugins/formkit/_index.scss" layer(components);
```

But whenever i try and import the utilities from tailwind like so:
```
@import "../theme/_index.scss" layer(theme);

@import "tailwindcss/preflight" layer(base);
@import "../base/_index.scss" layer(base);
@import "../plugins/postcss/_index.scss" layer(base);

@import "../plugins/formkit/_index.scss" layer(components);

@import "tailwindcss/utilities" layer(utilities);
```
The HMR stops working, it does still say that the HMR got updated but it wont actually apply the changes to the front-end.
Whenever i save my main scss file, the changes actually get applied.

**Comment by amiii123malviya** [comment]
Hello,

Is this issue still active?
I would like to work on this.

Best Regards.

**Comment by Matheun** [comment]
Yes, this is still an issue

**Comment by cris458carlo** [comment]
Hello,
you’re encountering an issue with Hot Module Replacement (HMR) when importing Tailwind CSS utilities. This can be frustrating, but let’s try to troubleshoot it together.

Here are a few things you can check:

Order of Imports: Ensure that the order of your imports is correct. Sometimes, the order can affect how styles are applied and how HMR works.
Tailwind Configuration: Verify that your tailwind.config.js is correctly set up and includes the necessary layers and paths.
Nuxt Configuration: Check your Nuxt configuration to ensure that it is properly set up to handle PostCSS and Tailwind CSS. Make sure you have the necessary modules and plugins installed and configured.
Cache Issues: Sometimes, caching can cause issues with HMR. Try clearing your cache or restarting your development server.
Dependencies: Ensure all your dependencies are up to date. Sometimes, mismatched versions can cause issues.
Error Logs: Check your browser’s console and terminal for any error messages that might give you more insight into what’s going wrong.
Here’s a sample configuration for Nuxt with Tailwind CSS and PostCSS that you can compare with your setup: // nuxt.config.js
export default {
  buildModules: [
    '@nuxt/postcss8',
    '@nuxtjs/tailwindcss',
  ],
  css: [
    '@/assets/css/main.scss',
  ],
  build: {
    postcss: {
      plugins: {
        tailwindcss: {},
        autoprefixer: {},
      },
    },
  },
}
If everything seems correct and you’re still facing issues, you might want to try a different [costcoess](https://www-costcoess.com) approach to importing Tailwind CSS utilities, such as using the @apply directive within your SCSS files instead of importing the utilities directly.

**Comment by Matheun** [comment]
Please read the issue before commenting, your talking about handling this issue for tailwind v3, im encountering this issue with tailwind v4.

Thus making your comment useless

**Comment by boylett** [comment]
Am experiencing this as well. Removing `@import "tailwindcss/utilities" layer(utilities);` from my root stylesheet fixes the problem but, of course, leaves me without any utilities.

Console logs:
```
[Error] TypeError: Importing a module script failed.
	warnFailedUpdate (client:174)
	(anonymous function) (client:213)
[Error] [hmr] Failed to reload /app/root.css. This could be due to syntax errors or importing non-existent modules. (see errors above)
	warnFailedUpdate (client:176)
	(anonymous function) (client:213)
```

**Comment by philipp-spiess** [comment]
Hey @Matheun  and @boylett! Does either of you have a minimal repro for this issue? We're currently working on rebuilding our Vite extension but haven't encountered this particular problem.

**Comment by boylett** [comment]
I'm using Vite and Remix. Frustratingly I can't seem to reproduce this on a fresh install - however, this does seem to be a widespread issue relegated only to Safari on macOS/iOS. Doesn't happen at all on Chrome or Firefox.

The same problem occurs when working with [SvelteKit](https://github.com/sveltejs/kit/issues/5208) as well as in [many](https://talk.observablehq.com/t/importing-javascript-now-fail-with-script-src-violations/4748/5) [other](https://anvil.works/forum/t/anvil-extras-local-storage-import-error-issue-on-mobile-device/19102) contexts, [many](https://laracasts.com/discuss/channels/vite/vite-hmr-broken-in-safari) [of](https://stackoverflow.com/q/76413373) [them](https://juejin.cn/post/7321930742401269797#heading-7) involving Safari. Note that doing a hard-refresh (`⌥+⌘+E` → `⌘R`) does indeed prevent the error until you next save a change.

Doesn't necessarily seem to be a Twilwind-specific issue, but possibly something to do with the way Safari [caches](https://www.reddit.com/r/reactjs/comments/zow6s0/do_you_have_any_bad_experiences_using_vite_for/) [modules](https://hospodarets.com/native-ecmascript-modules-dynamic-import#promise-api-caveats) constructed with frameworks like vite?

I opened Safari and Chrome side-by-side and went along developing as normal, and Chrome updated as expected with no errors at all while Safari failed miserably.

**tl;dr** – Likely a Safari caching issue.

---

Edit: Occasionally I am noticing that HMR mixes up the order of Preflight and Base/Utilities, causing preflight's [reset](https://github.com/tailwindlabs/tailwindcss/blob/ac6d4a6e8e9ae7ca197bf98d933ae2f205be3635/packages/tailwindcss/preflight.css#L7) to override all other utilities. Should I make a separate issue for this?

**Comment by philipp-spiess** [comment]
Thanks for the writeup @boylett. Agree that these issues you linked don't seem to be related to Tailwind.

The latest alpha release (`v4.0.0-alpha.23`) contains a lot of changes in regards to how the Vite plugin works. It would be great if both of you can try to upgrade and let us know if the issue persists? I wasn't able to repro this myself so any information on how to do that would be really helpful here.

**Comment by Matheun** [comment]
@philipp-spiess I manage to get it working now using the vite plugin and .css files. But i kindof really need to use scss files, for some reason they dont load correctly? i have no clue why or how.

When inspecting the style editor in the inspector, every variable is defined as expected and the layers are correct.
But when i add the class pt-4 to an element, nothing happens. The default variables are also not defined when inspecting the root html even tho they are in the style editor.

nuxt.config.ts
```
import tailwindcss from "@tailwindcss/vite";
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
    devtools: { enabled: true },
    
    css: ["./assets/scss/vendors/tailwind.scss"],
    
    vite: {
        plugins: [
            tailwindcss(),
        ],
    },
})
```

./assets/scss/vendors/tailwind.scss
```
@import "tailwindcss";
```

**Comment by Matheun** [comment]
# To recap:


## Case 1 - tailwindcss/vite
see: https://github.com/Matheun/tailwind-v4-case-1

`HMR` works correctly when using regular `.css` files and `<style>` tags.

This hover does result in me being forced to use `.css` files and regular `<style>` tags in my `.vue` files. So i won't be able to use `<style scoped lang="scss">` or `<style lang="scss">` tags in my `.vue` files. Or i won't be able to use `@apply` in `.vue` files.

Resulting in this setup being kindof useless when you need to use scss
```ts
// nuxt.config.ts
import tailwindcss from "@tailwindcss/vite";
export default defineNuxtConfig({
    css: [
        "./assets/style/tailwind.css",
        "./assets/style/main.scss"
    ],
    vite: {
        plugins: [
            tailwindcss(),
        ],
    },
})
```

```vue
// app.vue
<template>
  <div>
    <p class="pt-4">Test paragraph</p>
  </div>
</template>
<style>
@import 'tailwindcss/utilities'; 
@import 'tailwindcss/theme' theme(reference);
/* ^ having to declade this every time makes the project very prone to errors */

p {
  @apply pl-4;
}
</style>
```

## Case 2 - tailwindcss/vite
see: https://github.com/Matheun/tailwind-v4-case-2

Now vite doesnt load the scss files correctly - meaning the tailwind does get loaded but the dom somehow can find them/ apply the css variables
```ts
// nuxt.config.ts
import tailwindcss from "@tailwindcss/vite";
export default defineNuxtConfig({
    css: ["./assets/style/tailwind.scss"],
    vite: {
        plugins: [
            tailwindcss(),
        ],
    },
})
```

## Case 3 - tailwindcss/postcss
see: https://github.com/Matheun/tailwind-v4-case-3

Now i'm able to use scss and all the good stuff it provides, but `@apply` doesnt work in `.vue` files
```ts
// nuxt.config.ts
export default defineNuxtConfig({
    css: ["./assets/scss/vendors/tailwind.scss"],
    postcss: {
        plugins: {
            "@tailwindcss/postcss": {},
        },
    },
})
```

**Comment by philipp-spiess** [comment]
@Matheun Sorry for only getting back to you today, but yeah as you mentioned in a separat issue, this is related to the Vite plugin not supporting other pre-processors at the moment: https://github.com/tailwindlabs/tailwindcss/issues/14376

Regarding case 3: It seems unexpected that `.vue` files do not use the whole Post CSS pipeline when parsing style blocks. Have you tried using `<style lang="postcss">` tags?

**Comment by Matheun** [comment]
@philipp-spiess Just tried it, now i get the following error: 
`[postcss] Cannot apply unknown utility class: pl-4`

code:
```vue
<template>
  <div>
    <p class="pt-4">Test paragraph</p>
  </div>
</template>

<style lang="postcss">
p {
  @apply pl-4;
}
</style>
```

The following however does work, but is there a way around importing tailwind in every style stag? since we often have 75+ components and its a hassle debugging if we accidentally import tailwind the wrong way
```vue
<template>
  <div>
    <p class="pt-4">Test paragraph</p>
  </div>
</template>

<style lang="postcss">
@import 'tailwindcss/utilities'; 
@import 'tailwindcss/theme' theme(reference);
p {
  @apply pl-4;
}
</style>
```

Also works:
```vue
<template>
  <div>
    <p class="pt-4">Test paragraph</p>
  </div>
</template>

<style lang="scss">
@import 'tailwindcss/utilities'; 
@import 'tailwindcss/theme' theme(reference);
p {
  @apply pl-4;
}
</style>
```

**Comment by philipp-spiess** [comment]
Ah, yeah, the issue here is that the `pl` utility requires the `--spacing` `@theme` to be defined, otherwise we don't know what `4` should refer to, in this case.  It's enough to add _only_ the `@import 'tailwindcss/theme' theme(reference);` import in this case. 

Right now it's required to add that import unfortunately, since the CSS file is now our source of truth regarding the config. it's no longer one config for the whole project. You _can_ import your own `theme.css` file here instead via `theme(reference)` too, though, if you want to adjust the theme. 

We will have to think how we can make this easier, though. I'll get back to you about this.

**Comment by Matheun** [comment]
Allright, can see why thats the case.

And if i want to import a custom theme, i would do it like so?
```scss
  @import 'tailwindcss/theme' theme(url.scss);
```

Wouldn't the additionalData option in postcss be a possible fix for now? i can remember that i used this back in the day to load my sass variables globally

**Comment by Matheun** [comment]
And where can i read up on some more info about these options? i kept going through the release notes, but its kinda hard to get a good overview.

Also, where do you recommend i start/ what path/ which roadmap do i follow in doing my research into the source code of tailwind, since i want to start contributing and get a deeper understanding as to how the entire codebase works?

**Comment by philipp-spiess** [comment]
> And if i want to import a custom theme, i would do it like so?

That would be:

```css
@import './src/tailwind-theme.css' theme(reference);
```

The `theme(reference)` makes it so that this import will _not_ emit CSS variables for the theme, since you usually want those only once in your globals.

> Wouldn't the `additionalData` option in postcss be a possible fix for now? i can remember that i used this back in the day to load my sass variables globally

Ah, yep, maybe this works (although for relative imports I’m not sure it will use the right base path?). It does seem to only be a SCSS option though, so it won't work for all postcss setups.

> Also, where do you recommend i start/ what path/ which roadmap do i follow in doing my research into the source code of tailwind, since i want to start contributing and get a deeper understanding as to how the entire codebase works?

Yeah most of that info is only in PR descriptions right now, we haven't gotten around to writing docs yet, sorry! 😞

**Comment by Matheun** [comment]
I would like to setup some nuxt and vue examples so people have a base to get started, would this be of value/ usefull? i could maybe do a pull request into the playgrounds folder? or should i create a repo on my own github and get it linked somewhere?

Also, i would be down to get some documentation going? i'm currently diving into every v4 release and code.
Where should i then put the base of the doc?

**Comment by philipp-spiess** [comment]
> I would like to setup some nuxt and vue examples so people have a base to get started, would this be of value/ usefull? i could maybe do a pull request into the playgrounds folder? or should i create a repo on my own github and get it linked somewhere?

I think that would help but we don't want to add too many playgrounds in the repo since these quickly go outdated. There are some integration tests for both vue and nuxt though that we already maintain, which might be a good thing to look into:

- https://github.com/tailwindlabs/tailwindcss/blob/next/integrations/vite/nuxt.test.ts
- https://github.com/tailwindlabs/tailwindcss/blob/next/integrations/vite/vue.test.ts

I don't think we want to add one for the SCSS setup via postcss _just yet_, mostly because we're still thinking about wether we will make this setup easier. However one general test inside the `postcss` set of tests where we have the postcss processor first and then tailwind and ensure features work would actually be really helpful. So something like this one:

- https://github.com/tailwindlabs/tailwindcss/blob/next/integrations/postcss/index.test.ts#L4-L80

> Also, i would be down to get some documentation going? i'm currently diving into every v4 release and code.
Where should i then put the base of the doc?

Yea so the unfortunate answer is that we don't have a place for this yet. We just talked about docs in our sync yesterdan and it's likely going to be a fork off the v3 docs but with some larger changes planned. I don't think we know exactly what we want there yet so I don't think we can guarantee that a contribution there will be merged. I think timeline wise it'll be a few more weeks before we fully start this I'm afraid.

> Quick update on my HMR issue, i found the problem

I’m not sure I quite understand the issue yet. Maybe it helps if you share a repro here with the contents of your CSS file. Thank you!

**Comment by Matheun** [comment]
I'll setup some repo for you to have a look at.

Side question, is there any way to get @tailwindcss/typography working with postcss or do i need vite for that

**Comment by philipp-spiess** [comment]
@Matheun The plugin support should be independent of which client is used, so `@plugin '@tailwindcss/typography'` should work regardless of you using Vite or Postcss!

**Comment by Matheun** [comment]
> @Matheun The plugin support should be independent of which client is used, so `@plugin '@tailwindcss/typography'` should work regardless of you using Vite or Postcss!

Aaaaaah, yea i was doing it the vite way with a js file, didnt know is was this simple

**Comment by Matheun** [comment]
@philipp-spiess Updated the repo to show which import breaks the HMR and which works with HMR

https://github.com/Matheun/tailwind-v4-case-3

**Comment by Matheun** [comment]
@philipp-spiess ?

**Comment by Matheun** [comment]
Also, When i use the following code, the HMR for that file will not work when i save the file instead of the main scss
```scss
@import "tailwindcss/preflight" layer(base);
@import "../../../components/button/button.scss";
```

This makes the HMR work
```scss
@import "tailwindcss/preflight" layer(base);
@import "../../../components/button/button.scss" layer(components);
```

**Comment by philipp-spiess** [comment]
@Matheun Hey! Thank you for reporting. There indeed was something wrong with CSS file dependencies not being watched. We have a PR to fix this here: https://github.com/tailwindlabs/tailwindcss/pull/14561

**Comment by RobinMalfait** [comment]
Hey!

This should be fixed by #14561, and will be available in the next alpha release.

**Comment by philipp-spiess** [comment]
@Matheun Hey! Responding here since you said you still have issues with HMR and this setup. I was looking into it a bit and what happens here seems really funny. I noticed in the logs of Vite that if you have the syntax _without_ the `layer()` modifier, the SCSS extension is actually inlining the `@import`. When this happens, it doesn't tell Vite that it now has a new dependency so changes to the `.scss` files are never picked up. The custom `layer()` works because it will bypass the SCSS import rules leaving it to Tailwind to do the imports where we do properly track these.

A workaround for you is to not rely on the SCSS extension built-into Nuxt but instead put `postcss-import` into the stack manually. This will resolve imports correctly.

Our official recommendation though is to not use SCSS at all if you use Tailwind CSS 😄 Curious what kind of features you are missing?
